assets/php/class-wgpb-block-grid-base.php

```diff
@@ -0,0 +1,340 @@
+		$query_args['tax_query']      = array(); // phpcs:ignore WordPress.DB.SlowDBQuery
+
+		$this->set_block_query_args( $query_args );
+
+		$ordering_args         = WC()->query->get_catalog_ordering_args( $query_args['orderby'], $query_args['order'] );
+		$query_args['orderby'] = $ordering_args['orderby'];
+		$query_args['order']   = $ordering_args['order'];
+		if ( $ordering_args['meta_key'] ) {
+			$query_args['meta_key'] = $ordering_args['meta_key']; // phpcs:ignore WordPress.DB.SlowDBQuery
+		}
+
+		// Categories.
+		$this->set_categories_query_args( $query_args );
+
+		// Always query only IDs.
+		$query_args['fields'] = 'ids';
+
+		return $query_args;
+	}
+
+	/**
+	 * Set args specific to this block
+	 *
+	 * @param array $query_args Query args.
+	 */
+	abstract protected function set_block_query_args( &$query_args );
+
+	/**
+	 * Set categories query args.
+	 *
+	 * @param array $query_args Query args.
+	 */
+	protected function set_categories_query_args( &$query_args ) {
+		if ( ! empty( $this->attributes['categories'] ) ) {
+			$categories = array_map( 'absint', $this->attributes['categories'] );
+
+			$query_args['tax_query'][] = array(
+				'taxonomy'         => 'product_cat',
+				'terms'            => $categories,
+				'field'            => 'term_id',
+				'operator'         => 'all' === $this->attributes['catOperator'] ? 'AND' : 'IN',
+
+				/*
+				 * When cat_operator is AND, the children categories should be excluded,
+				 * as only products belonging to all the children categories would be selected.
+				 */
+				'include_children' => 'all' === $this->attributes['catOperator'] ? false : true,
+			);
+		}
+	}
+
+	/**
+	 * Run the query and return an array of product IDs
+	 *
+	 * @return array List of product IDs
+	 */
+	protected function get_products() {
+		// @todo Cache this
+		if ( 'product-top-rated' === $this->block_name ) {
+			add_filter( 'posts_clauses', array( __CLASS__, 'order_by_rating_post_clauses' ) );
+			$query = new WP_Query( $this->query_args );
+			remove_filter( 'posts_clauses', array( __CLASS__, 'order_by_rating_post_clauses' ) );
+		} else {
+			$query = new WP_Query( $this->query_args );
+		}
+
+		$results = wp_parse_id_list( $query->posts );
+
+		// Remove ordering query arguments which may have been added by get_catalog_ordering_args.
+		WC()->query->remove_ordering_args();
+
+		return $results;
+	}
+
+	/**
+	 * Render the Products block.
+	 *
+	 * @return string Rendered block type output.
+	 */
+	public function render() {
+		$products = $this->get_products();
+		$classes  = $this->get_classes();
+
+		$output = implode( '', array_map( array( $this, 'render_product' ), $products ) );
+
+		return sprintf( '<ul class="%s">%s</ul>', $classes, $output );
+	}
+
+	/**
+	 * Get the list of classes to apply to this block.
+	 *
+	 * @return string space-separated list of classes.
+	 */
+	public function get_classes() {
```

Comment:
Minor nitpick. Might add clarity to specify this is a CSS class and what it's for since we're rendering multiple things. e.g. `get_product_list_class()` or similar.